### PR TITLE
Enable AOT perf support

### PR DESCRIPTION
**Pull request by lum1n0us**
- use `wamrc --perf-profile` to generate .aot contains fp operations
- use `iwasm --perf-profile` to create perf map for `perf record`

### CODE REVIEW

File: core/iwasm/common/wasm_runtime_common.c [hunk 1]
```diff
@@ -6148,3 +6152,19 @@ wasm_runtime_get_context(WASMModuleInstanceCommon *inst, void *key)
     return wasm_native_get_context(inst, key);
 }
 #endif /* WASM_ENABLE_MODULE_INST_CONTEXT != 0 */
+
+#if WASM_ENABLE_AOT != 0
```
Comment (lum1n0us) [inline core/iwasm/common/wasm_runtime_common.c:6156]: Is it better if merge JIT perf support operations here? w/o using a member in `LLVMJITOptions`?

Comment (wenyongh) [reply core/iwasm/common/wasm_runtime_common.c:6156]: Yes, had better just merge them into one option here. And had better introduce a cmake variable (e.g. WAMR_BUILD_LINUX_PERF_SUPPORT) and compiler macro (e.g. WASM_ENABLE_LINUX_PERF_SUPPORT) for this feature. Make it configurable so as not to increase footprint when it isn't enabled.

File: product-mini/platforms/posix/main.c [hunk 2]
```diff
@@ -67,6 +66,9 @@ print_help()
     printf("                           Use comma to separate, e.g. --enable-segue=i32.load,i64.store\n");
     printf("                           and --enable-segue means all flags are added.\n");
 #endif
+#endif /* WASM_ENABLE_JIT != 0*/
+#if WASM_ENABLE_AOT != 0 || WASM_ENABLE_JIT != 0
+    printf("  --perf-profile           Enable linux perf support. It works in aot and llvm-jit.\n");
```
Comment (wenyongh) [inline product-mini/platforms/posix/main.c:71]: Is it better to use `--enable-perf-profile`?

File: wamr-compiler/main.c [hunk 3]
```diff
@@ -526,6 +527,9 @@ main(int argc, char *argv[])
             native_lib_list[native_lib_count++] = argv[0] + 13;
         }
 #endif
+        else if (!strncmp(argv[0], "--perf-profile", 14)) {
```
Comment (wenyongh) [inline wamr-compiler/main.c:530]: Is it better to use `--enable-perf-profile`?

File: product-mini/platforms/posix/main.c (hunk 2 above)
Comment (lum1n0us) [reply product-mini/platforms/posix/main.c:71]: how about `--support-perf` or `--enable-perf`? Don't want to make it longer

File: core/iwasm/aot/aot_loader.c [hunk 4]
```diff
@@ -2764,6 +2764,57 @@ load_relocation_section(const uint8 *buf, const uint8 *buf_end,
...
+{
+    uint32 func_sz;
+    if (func_idx_adj + 1 < module->func_count) {
+        func_sz = module->func_ptrs[func_idx_adj + 1]
+                  - module->func_ptrs[func_idx_adj];
+    }
+    else {
+        func_sz = module->code_size
+                  - (module->func_ptrs[func_idx_adj] - module->code);
+    }
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:2785]: An issue is that the function pointers may be not ordered, e.g. when AOT static PGO is enabled. Had better traverse all the pointers and get a pointer which is larger than current pointer and is the smallest.

File: core/iwasm/include/wasm_c_api.h [hunk 5]
```diff
@@ -181,24 +181,28 @@ typedef union MemAllocOption {
 struct wasm_config_t {
     mem_alloc_type_t mem_alloc_type;
     MemAllocOption mem_alloc_option;
-    bool linux_perf_support;
+#if WASM_ENABLE_LINUX_PERF != 0
```
Comment (wenyongh) [inline core/iwasm/include/wasm_c_api.h:184]: Don't add macro here, like RuntimeInitArgs in wasm_export.h, the embedder may use the header file and libiwasm.so/libvmlib.a in his project, and doesn't define the macro to compile WAMR again.

File: core/iwasm/include/wasm_c_api.h [hunk 6]
```diff
@@ -181,24 +181,28 @@ typedef union MemAllocOption {
...
- * - linux_perf_support is false
+ * - enable_linux_perf is false
  */
 WASM_API_EXTERN own wasm_config_t* wasm_config_new(void);
 
 // Embedders may provide custom functions for manipulating configs.
 WASM_API_EXTERN own wasm_config_t*
 wasm_config_set_mem_alloc_opt(wasm_config_t *, mem_alloc_type_t, MemAllocOption *);
 
+#if WASM_ENABLE_LINUX_PERF != 0
```
Comment (wenyongh) [inline core/iwasm/include/wasm_c_api.h:202]: Same as above, had better remove the macro and add comment that this API requires building WAMR with linux perf enabled.

File: wamr-compiler/CMakeLists.txt [hunk 7]
```diff
@@ -49,7 +49,12 @@ add_definitions(-DWASM_ENABLE_MODULE_INST_CONTEXT=1)
 
 if (WAMR_BUILD_LLVM_LEGACY_PM EQUAL 1)
   add_definitions(-DWASM_ENABLE_LLVM_LEGACY_PM=1)
-endif()
+endif ()
+
+if (WAMR_BUILD_LINUX_PERF EQUAL 1)
+  message ("-- Enable linux perf support")
+  add_definitions(-DWASM_ENABLE_LINUX_PERF=1)
+endif ()
```
Comment (wenyongh) [inline wamr-compiler/CMakeLists.txt:57]: Had better add the macro directly (remove if/endif), and just using `wamrc --enable-linux-perf` to enable the feature for AOT file. Or we need to compile wamrc individually.

File: wamr-compiler/main.c [hunk 8]
```diff
@@ -325,6 +328,9 @@ main(int argc, char *argv[])
     void *native_handle_list[8] = { NULL };
     uint32 native_handle_count = 0;
 #endif
+#if WASM_ENABLE_LINUX_PERF != 0
```
Comment (wenyongh) [inline wamr-compiler/main.c:331]: Same as above, remove macro controls in this file.

File: core/iwasm/aot/aot_loader.c [hunk 9]
```diff
@@ -10,6 +10,7 @@
 #include "../common/wasm_runtime_common.h"
 #include "../common/wasm_native.h"
 #include "../compilation/aot.h"
+#include <stdint.h>
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:13]: Had better remove this line, the platform_internal.h should already include related header file?

File: core/iwasm/aot/aot_loader.c [hunk 10]
```diff
@@ -2764,6 +2765,97 @@ load_relocation_section(const uint8 *buf, const uint8 *buf_end,
...
+static int
+compare_func_ptrs(const void *fp1, const void *fp2)
+{
+    return (intptr_t)(*(void **)fp1) - (intptr_t)(*(void **)fp2);
+}
+
+static void *
+sort_func_ptrs(const AOTModule *module, char *error_buf, uint32 error_buf_size)
+{
+    size_t content_len = sizeof(void *) * module->func_count;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:2800]: Had better `uint64 content_len = (uint64)sizeof(void *) * module->func_count`

File: core/iwasm/aot/aot_loader.c [hunk 11]
```diff
@@ -2764,6 +2765,97 @@ load_relocation_section(const uint8 *buf, const uint8 *buf_end,
...
+    void *sorted_func_ptrs =
+        loader_malloc(content_len, error_buf, error_buf_size);
+    if (!sorted_func_ptrs)
+        return NULL;
+
+    b_memcpy_s(sorted_func_ptrs, content_len, module->func_ptrs, content_len);
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:2806]: Had better use bh_memcpy_s, we normally use that, it adds more checks

File: core/iwasm/aot/aot_loader.c [hunk 12]
```diff
@@ -2764,6 +2765,97 @@ load_relocation_section(const uint8 *buf, const uint8 *buf_end,
...
+        LOG_WARNING("can't create /tmp/perf-%d.map, because %s", pid,
+                    strerror(errno));
+        goto quit;
+    }
+
+    for (i = 0; i < module->func_count; i++) {
+        memset(perf_map_info, 0, 128);
+        snprintf(perf_map_info, 128, "%lx  %x  aot_func#%u\n",
+                 get_func_addr(module, sorted_func_ptrs, i),
+                 get_func_size(module, sorted_func_ptrs, i), i);
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:2840]: get_func_addr(.., i) returns sorted_func_ptrs[i], but it seems that we can not ensure that after sorted, `aot_func#i`'s func_ptr is sorted_func_ptrs[i].
Had better add an array to save the func indexes of each sorted func_ptr?

File: core/iwasm/common/wasm_runtime_common.c [hunk 13]
```diff
@@ -662,14 +662,17 @@ wasm_runtime_full_init(RuntimeInitArgs *init_args)
...
     llvm_jit_options.segue_flags = init_args->segue_flags;
-    llvm_jit_options.linux_perf_support = init_args->linux_perf_support;
+#endif
+
+#if WASM_ENABLE_LINUX_PERF != 0
+    wasm_runtime_set_linux_perf(init_args->enable_linux_perf);
+#else
+    if (init_args->enable_linux_perf)
+        LOG_WARNING("in order to enable linux perf support, need to recompile "
+                    "with -DWAMR_BUILD_LINUX_PERF=1");
```
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:675]: how about "warning: to enable linux perf support, please recompile with `-DWAMR_BUILD_LINUX=1`"?

File: core/iwasm/aot/aot_loader.c [hunk 14]
```diff
@@ -2764,6 +2765,97 @@ load_relocation_section(const uint8 *buf, const uint8 *buf_end,
...
+    bool ret = false;
+
+    sorted_func_ptrs = sort_func_ptrs(module, error_buf, error_buf_size);
+    if (!sorted_func_ptrs)
+        goto quit;
+
+    snprintf(perf_map_info, 128, "/tmp/perf-%d.map", pid);
+    perf_map = fopen(perf_map_info, "w");
+    if (!perf_map) {
+        LOG_WARNING("can't create /tmp/perf-%d.map, because %s", pid,
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:2831]: Had better add "warning: "

File: wamr-compiler/main.c [hunk 15]
```diff
@@ -325,6 +326,7 @@ main(int argc, char *argv[])
     void *native_handle_list[8] = { NULL };
     uint32 native_handle_count = 0;
 #endif
+    bool enable_enable_linux_perf = false;
```
Comment (wenyongh) [inline wamr-compiler/main.c:329]: Had better `enable_linux_perf`?

File: wamr-compiler/main.c [hunk 16]
```diff
@@ -184,10 +184,11 @@ print_help()
...
     printf("                                --emit-custom-sections=section1,section2,sectionN\n");
 #if BH_HAS_DLFCN
-    printf("  --native-lib=<lib>       Register native libraries to the WASM module, which\n");
-    printf("                           are shared object (.so) files, for example:\n");
-    printf("                             --native-lib=test1.so --native-lib=test2.so\n");
+    printf("  --native-lib=<lib>        Register native libraries to the WASM module, which\n");
+    printf("                            are shared object (.so) files, for example:\n");
+    printf("                              --native-lib=test1.so --native-lib=test2.so\n");
 #endif
+    printf("  --enable-linux-perf       Enable linux perf support\n");
```
Comment (wenyongh) [inline wamr-compiler/main.c:191]: Should we add WASM_ENABLE_LINUX_PERF control for this file?

File: core/iwasm/aot/aot_loader.c [hunk 17]
```diff
@@ -2764,6 +2764,104 @@ load_relocation_section(const uint8 *buf, const uint8 *buf_end,
...
+                    strerror(errno));
+        goto quit;
+    }
+
+    for (i = 0; i < module->func_count; i++) {
+        memset(perf_map_info, 0, 128);
+        snprintf(perf_map_info, 128, "%lx  %x  aot_func#%u\n",
+                 (uintptr_t)sorted_func_ptrs[i].ptr,
+                 get_func_size(module, sorted_func_ptrs, i),
+                 sorted_func_ptrs[i].idx);
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:2846]: Seems getting func ptr/size of aot_func#i is invalid? Should be like below?
```C
    /* Get func ptr of aot_func#i */
    for (j = 0; j < module->func_count; j++) {
        if (sorted_func_ptrs[j].idx == i)
            return sorted_func_ptrs[j].ptr;
    }

    /* Get func size of aot_func#i */
    for (j = 0; j < module->func_count; j++) {
        if (sorted_func_ptrs[j].idx == i) {
            if (j < module->func_count - 1)
               return sorted_func_ptrs[j + 1].ptr - sorted_func_ptrs[j].ptr;
            else
               return (uintptr_t)module->code + module->code_size
                          - (uintptr_t)sorted_func_ptrs[j].ptr;
        }
    }
```

Comment (lum1n0us) [reply core/iwasm/aot/aot_loader.c:2846]: planning to make a traversal in the sorted order, from ptrs with lower addresses to with high addresses.

**Pull request merged**